COSE envelope signature spec updates

SteveLasker opened the pull request:
Tracking the merger of the [cose-envelope](https://github.com/notaryproject/notaryproject/tree/cose-envelope) branch into main.

shizhMSFT committed:
COSE Signature Envelope (#139)

* cbor payload
* cose envelope
* finalize payload
* remove JWS
* refine notations
* refine datetime

Signed-off-by: Shiwei Zhang <shizh@microsoft.com>

Merging into the remote-siging branch for further iteration, addressing the PR feedback with additional PRs into this branch.

SteveLasker committed:
Support for JWT and COSE envelopes (#145)

Signed-off-by: Steve Lasker <stevenlasker@hotmail.com>

gokarnm committed:
Updates to signature specification (#147)

Updates to signature specification
* Moves signed attributes out of payload
* Defines new signed attributes and marks them as critical/informational
* Introduces signature scheme to support different trust models

Signed-off-by: Milind Gokarn <gokarnm@amazon.com>

iamsamirzon commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,134 @@
+In COSE ([rfc8152](https://datatracker.ietf.org/doc/html/rfc8152)), data is stored as either payload or headers (protected and unprotected).
+Notary v2 supports [COSE_Sign1_Tagged](https://datatracker.ietf.org/doc/html/rfc8152#section-4.2) as a signature envelope with some additional constraints on the header fields.
+
+A JWS based signature will be persisted with a blob `mediaType` of `"application/cose"`
```
Typo: I believe this should be COSE instead of JWS on this line

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+
+The payload contains the subject manifest and other attributes that have to be integrity protected.
+
+- **`subject`**(*descriptor*): A REQUIRED top-level node consisting of the manifest that needs to be integrity protected.
```
Inconsistent with above, which says `"notary"`.

letmaik replied:
What is a "node"? Probably this should be "property"?

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+Also, there shouldn’t be any additional fields other than ones specified in JWSPayload, ProtectedHeader, and UnprotectedHeader.
+
+**JWSPayload a.k.a. Claims**:
+Notary v2 is using one private claim (`notary`) and two public claims (`iat` and `exp`).
```
How is "public" and "private" defined? Isn't this a notion of JWTs?

dtzar replied:
@letmaik There are no changes to signature-envelope-jws.md at this point, so marking all the ones for this file as resolved.

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+- **`iat`**(*number*): The REQUIRED property Issued-at(`iat`) identifies the time at which the signature was issued.
+- **`exp`**(*number*): This OPTIONAL property contains the expiration time on or after which the signature must not be considered valid.
+
+To leverage JWS claims validation functionality already provided by libraries, we have defined `iat`, `exp` as top-level nodes.
```
Again, "nodes" -> "properties".

letmaik replied:
Which libraries? JWS libraries or JWT libraries?

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+  This will be used to version different variations of JWS signature.
+  The supported value is `application/vnd.cncf.notary.v2.jws.v1`.
+- **`crit`**(*array of strings*): This REQUIRED property lists the headers that implementation MUST understand and process.
+  The value MUST be `["cty"]`.
```
This makes no sense. `"cty"` is a standard JOSE header parameter and hence mustn't be included in `"crit"`. The spec says:
> Producers MUST NOT include Header Parameter names defined by this specification [...] Recipients MAY consider the JWS to be invalid if the critical list contains any Header Parameter names defined by this specification.

letmaik commented on signature-envelope-jws.md:
````diff
@@ -0,0 +1,142 @@
+
+```jsonc
+{
+    "alg": "RS256",
````
Example `"alg"` doesn't follow own restrictions declared later on. Should be `"PS256"`.

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+- **`x5c`** (*array of strings*): This REQUIRED property contains the list of X.509 certificate or certificate chain([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign the JWS.
+  The certificate containing the public key corresponding to the key used to digitally sign the JWS MUST be the first certificate.
+
+- **`timestamp`** (*string*): This OPTIONAL property is used to store time stamp token.
```
Duplicated.

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+
+- **`timestamp`** (*string*): This OPTIONAL property is used to store time stamp token.
+  Only [RFC3161]([rfc3161](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2)) compliant TimeStampToken are supported.
+- **`x5c`** (*array of strings*): This REQUIRED property contains the list of X.509 certificate or certificate chain([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign the JWS.
```
Duplicated.

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+
+- **`timestamp`** (*string*): This OPTIONAL property is used to store time stamp token.
+  Only [RFC3161]([rfc3161](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2)) compliant TimeStampToken are supported.
+- **`x5c`** (*array of strings*): This REQUIRED property contains the list of X.509 certificate or certificate chain([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign the JWS.
```
@briankr-ms Just a ping, in case you had further comments to the earlier discussion on putting x5c in protected vs unprotected. See https://github.com/notaryproject/notaryproject/pull/139#discussion_r824276018.

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+        "size": 16724
+    },
+    "annotations": {
+        "io.cncf.notary.x509certs.fingerprint.sha256": "[\"B7A69A70992AE4F9FF103EBE04A2C3BA6C777E439253CE36562E6E98375068C3\" \"932EB6F5598435D4EF23F97B0B5ACB515FAE2B8D8FAC046AB813DDC419DD5E89\"]"
```
Inconsistent with `signature-specification.md`: `io.cncf.notary.x509.fingerprint.sha256`

letmaik commented on signature-specification.md:
```diff
@@ -1,256 +1,125 @@
   It is being used to store information about the signature.
   Keys using the `io.cncf.notary` namespace are reserved for use in Notary and MUST NOT be used by other specifications.
-  - **`io.cncf.notary.x509certs.fingerprint.sha256`**: A REQUIRED annotation whose value contains the list of SHA-256 fingerprint of signing certificate and certificate chain used for signature generation.
+  - **`io.cncf.notary.x509.fingerprint.sha256`**: A REQUIRED annotation whose value contains the list of SHA-256 fingerprint of signing certificate and certificate chain used for signature generation.
```
Not clear how the fingerprints are computed exactly.

letmaik commented on signature-specification.md:
````diff
@@ -1,256 +1,125 @@
-    "signature": "Base64Url( sign( ASCII( <Base64Url(ProtectedHeader)>.<Base64Url(JWSPayload)> )))"  
-}
-```
+- Any claims MUST NOT be stored/appended in the payload itself, as the payload is only parsed and processed once the signature has been verified (signature is valid, and from a trusted key) and trust is established.
````
Wouldn't `"exp"` also fall in this category? In `signature-envelope-jws.md` it's part of the payload, right?

letmaik commented on signature-specification.md:
````diff
@@ -1,256 +1,125 @@
-```
+- Any claims MUST NOT be stored/appended in the payload itself, as the payload is only parsed and processed once the signature has been verified (signature is valid, and from a trusted key) and trust is established.
+- Specific claims can be either required or optional.
+- Claims that if present, MUST be processed by a verifier MUST be marked as critical.
````
I don't understand this requirement. It seems that this excludes certain envelope formats, but see also my other comment in -jws.md about how `"crit": ["cty"]` is going against the JWS spec.

letmaik commented on signature-specification.md:
```diff
@@ -1,256 +1,125 @@
+#### Standard attributes
 
-## Signature Algorithm Requirements
+- **Notary Signature scheme** (critical): The trust model used by Notary currently supports the following values for this claim `notary.signer.x509` and `notary.signingservice.x509`. Both signature schemes use the same set of claims defined in this document. One difference is that `notary.signer.x509` uses a TSA signature to determine authenticated signing time (timestamp), and `notary.signingser ...
```
I couldn't find this claim in the JWS md for example. Shouldn't it be there?

letmaik commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,134 @@
+  [
+    / protected / << {
+      / crit / 2: [
+        / cty / 3,
```
COSE spec says "Integer labels in the range of 0 to 7 SHOULD be omitted.". This is less strict than JWS which forbids it, but would probably do the same for consistency.

letmaik commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,134 @@
+- **`crit`** (*array of integers or strings*): This REQUIRED property (label: `2`) lists the headers that implementation MUST understand and process.
+  The array MUST contain `3` (`cty`), and `signingtime`. If `expiry` is presented, the array MUST also contain `expiry`.
+- **`cty`** (*string*): The REQUIRED property content-type (label: `3`) is used to declare the media type of the secured content (the payload).
+- **`signingtime`** (*datetime*): The REQUIRED property identifies the time at which the signature was generated.
```
Should be more specific on the format. `*datetime*` isn't defined anywhere.

SteveLasker replied:
@letmaik, which RFC should we reference here?

BryceDFisher replied:
RFC 7519 specifies this field as number of _seconds_ since epoch, if it's to be aligned with formats used in JWT.

dtzar replied:
Its value is the number of seconds from `1970-01-01T00:00Z` in UTC time, commonly known as UNIX timestamp

letmaik commented on signature-envelope-jws.md:
```diff
@@ -0,0 +1,142 @@
+
+- **`subject`**(*descriptor*): A REQUIRED top-level node consisting of the manifest that needs to be integrity protected.
+  Please refer [Payload](#payload) section for more details.
+- **`iat`**(*number*): The REQUIRED property Issued-at(`iat`) identifies the time at which the signature was issued.
```
Needs more details on the format, i.e. UNIX timestamp in some unit. Same for `exp`.

BryceDFisher replied:
RFC 7519 specifies these as _seconds_ since epoch

dtzar replied:
jws not in scope for this PR

letmaik commented on signature-envelope-cose.md:
````diff
@@ -0,0 +1,134 @@
+    ]
+    ```
+3. Encode `Sig_structure` into a CBOR object as a byte string named `ToBeSigned`.
+4. Compute the signature on the `ToBeSigned` constructed in the previous step by using the signature algorithm defined in the corresponding header element: `alg`.
````
`alg` is missing in the protected headers above.

dtzar replied:
This is present in the spec

letmaik commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,134 @@
+
+Note: `<<` and `>>` are used to notate the byte string resulting from encoding the data item.
+
+- **`timestamp`** (*byte string*): This OPTIONAL property is used to store time stamp token.
```
Should be more specific on format, i.e. DER encoding.

SteveLasker replied:
Can you please propose what you'd like?

dtzar replied:
Its value is the number of seconds from `1970-01-01T00:00Z` in UTC time, commonly known as UNIX timestamp

letmaik commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,134 @@
+- **`crit`** (*array of integers or strings*): This REQUIRED property (label: `2`) lists the headers that implementation MUST understand and process.
+  The array MUST contain `3` (`cty`), and `signingtime`. If `expiry` is presented, the array MUST also contain `expiry`.
+- **`cty`** (*string*): The REQUIRED property content-type (label: `3`) is used to declare the media type of the secured content (the payload).
+- **`signingtime`** (*datetime*): The REQUIRED property identifies the time at which the signature was generated.
```
Any reason this can't use the JWT name `"iat"`? Same for expiry `"exp"`.

SteveLasker replied:
Happy to change to whatever folks prefer

BryceDFisher replied:
Is this the same as the difference between iat and nbf?  The signature could be generated at one time but not be considered valid until its issued?

dtzar replied:
We're consistently using io.cncf.notary.expiry on both and io.cncf.notary.signingTime

shizhMSFT committed:
Update COSE spec (#179)

* Update signature-envelope-cose.md (#153)

The `crit` header should be optional, not required. The spec already defines `cty` and `signingtime` as required, so a compliant verifier should reject any signature that does not include them, irrespective of whether or not the envelope also declares them as critical.

An appropriate use of `crit` would be if a signer intends to include a new header not defined in the spec, to instruct verifiers that they must understand and process that header in order for the signature to be successfully validated. Another scenario is if a signer intends to mandate processing of an otherwise optional header (e.g., `expiry`).

In other words, `crit` can be used by a signer to add requirements to the verification flow, but it is unnecessary and redundant if the verifier already has the same requirements.

Signed-off-by: Brian Krell <briankr@microsoft.com>

* `cose-envelope`: merge changes from `main` (#177)

* Align COSE spec with JWS

Signed-off-by: letmaik <maik.riechert@microsoft.com>
Co-authored-by: Steve Lasker <stevenlasker@hotmail.com>
Co-authored-by: Quim Muntal <quimmuntal@gmail.com>
Co-authored-by: Quim Muntal <qmuntaldiaz@microsoft.com>
Co-authored-by: Milind Gokarn <milind81@gmail.com>
Co-authored-by: Milind Gokarn <gokarnm@amazon.com>
Co-authored-by: Shiwei Zhang <shizh@microsoft.com>
Co-authored-by: Pritesh Bandi <priteshbandi@gmail.com>
Co-authored-by: Sajay Antony <1821104+sajayantony@users.noreply.github.com>

* update CBOR spec

Co-authored-by: Steve Lasker <stevenlasker@hotmail.com>
Co-authored-by: qmuntal <qmuntaldiaz@microsoft.com>
Co-authored-by: Milind Gokarn <gokarnm@amazon.com>
Co-authored-by: Pritesh Bandi <pritesb@amazon.com>
Co-authored-by: Sajay Antony <sajaya@microsoft.com>
Signed-off-by: Shiwei Zhang <shizh@microsoft.com>

Co-authored-by: Brian Krell <57008892+briankr-ms@users.noreply.github.com>
Co-authored-by: Maik Riechert <maik.riechert@arcor.de>
Co-authored-by: Steve Lasker <stevenlasker@hotmail.com>
Co-authored-by: Quim Muntal <quimmuntal@gmail.com>
Co-authored-by: Quim Muntal <qmuntaldiaz@microsoft.com>
Co-authored-by: Milind Gokarn <milind81@gmail.com>
Co-authored-by: Milind Gokarn <gokarnm@amazon.com>
Co-authored-by: Pritesh Bandi <priteshbandi@gmail.com>
Co-authored-by: Sajay Antony <1821104+sajayantony@users.noreply.github.com>
Co-authored-by: Pritesh Bandi <pritesb@amazon.com>
Co-authored-by: Sajay Antony <sajaya@microsoft.com>

dtzar committed:
Merge branch 'main' into cose-envelope

Commit:
Merge branch 'cose-envelope' of github.com:notaryproject/notaryproject into cose-envelope

thomas-fossati commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+
+## Storage
+
+A COSE signature envelope will be stored in an OCI registry as a blob, and referenced in the signature manifest as a blob with `mediaType` of `"application/cose"`.
```
a more precise media type to use in this context would be `application/cose; cose-type="cose-sign1"` (see [Section 2 of STD 96](https://www.rfc-editor.org/rfc/rfc9052#section-2)).

letmaik replied:
I think this is redundant since the COSE envelopes are tagged anyway. In my mind this would make client code harder to write because now you need to define how media type parameters are encoded in this context. This is not HTTP (which has its own definition for encoding those).

thomas-fossati replied:
> I think this is redundant since the COSE envelopes are tagged anyway.

Yeah it's redundant.  Whether this hint is worthwhile depends on how processing is implemented.  With the extra information, the COSE payload can be dispatched directly to the right API, whilst without, the implementation needs to peek into the COSE tag to dispatch.

> In my mind this would make client code harder to write because now you need to define how media type parameters are encoded in this context. This is not HTTP (which has its own definition for encoding those).

Not sure I understand what the encoding problem is?  The media parameters that'd be used here are all UTF-8 strings (which both JSON and I-JSON support off-the-shelf) with a-priori known values.

( I have the feeling I am missing something though :-) )

letmaik replied:
In practice, even outside the context of Email or HTTP headers, people use the same syntax as first defined in https://datatracker.ietf.org/doc/html/rfc2045#section-5.1 to encode parameters, especially for protocols/formats where you want to use a single string to encode the media type. 
But it has to be made explicit with a reference that this is what's happening here.
From an implementation point of view, for example in Python, clients would have to do the following to match the media type (note the `parse_header` function name):
```py
import cgi
mediatype, params = cgi.parse_header('application/cose; cose-type="cose-sign1"')
if mediatype == "application/cose" and params["cose-type"] == "cose-sign1":
  # process
```
This is necessary since whitespace becomes a factor now. Previously, a simple string comparison would have been enough: `mediatype == "application/cose"`.
Of course you could prescribe that the media type string must always use a specific fixed string that can be matched by string comparison. There is some chance that people will mess it up though.

I think, given that the ORAS artifact type already says that it is a (single) signature (`"artifactType": "application/vnd.cncf.notary.signature"`), I don't see why it has to be repeated here using the parameter.

dtzar replied:
Agree with letmaik - going to close this comment out.

thomas-fossati commented on signature-envelope-cose.md:
````diff
@@ -0,0 +1,197 @@
+}
+```
+
+Note: `<<` and `>>` are used to notate the byte string resulting from encoding the data item.
````
```suggestion
Note: `<<` and `>>` are used to notate the CBOR byte string resulting from encoding the data item.
```

thomas-fossati commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+
+Note: `<<` and `>>` are used to notate the byte string resulting from encoding the data item.
+
+- **[`x5chain`](https://datatracker.ietf.org/doc/html/draft-ietf-cose-x509-08#section-2)** (*array of bstr*): This REQUIRED parameter (label `33` by [IANA](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)) contains the ordered list of X.509 certificate or certificate chain ([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign ...
```
```suggestion
- **[`x5chain`](https://datatracker.ietf.org/doc/html/draft-ietf-cose-x509-08#section-2)** (*array of bstr*): This REQUIRED parameter (label `33` by [IANA](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)) contains the ordered list of X.509 certificate or certificate chain ([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign the COSE. The certificate chain is represented as an array of certificate, each certificate in the array is DER encoded and then wrapped in a CBOR byte string. The certificate containing the public key corresponding to the key used to digitally sign the COSE MUST be the first certificate, followed by the intermediate and root certificates in the correct order. Refer [*Certificate Chain* unsigned attribute](signature-specification.md#unsigned-attributes) for more details. Optionally, this header can be presented in the protected header.
```

thomas-fossati commented on signature-specification.md:
```diff
@@ -43,6 +43,7 @@ Signature Manifest Example
 - **`blobs`** (*array of objects*): This REQUIRED property contains collection of only one [artifact descriptor][artifact-descriptor] referencing signature envelope.
   - **`mediaType`** (*string*): This REQUIRED property contains media type of signature envelope blob. Following values are supported
     - `application/jose+json`
+    - `application/cose`
```
similar to the suggestion I made above, I think this should be paired by the right `cose-type` parameter.

dtzar replied:
[same as above](https://github.com/notaryproject/notaryproject/pull/148/files#r975829734), redundant/unnecessary

thomas-fossati commented on signature-specification.md:
```diff
@@ -165,6 +168,127 @@ These attributes are considered unsigned with respect to the signing key that ge
+
+**Implementation Constraints**: Notary v2 implementation MUST enforce the following constraints on signature generation and verification:
+
+1. `alg` header value MUST NOT be `none` or any symmetric-key algorithm such as `HMAC`.
```
re: prohibition on HMAC.  This may be redundant since I don't think one can use symmetric key algorithms with COSE Sign/Sign1.  Those should be only available for Mac or Mac0 COSEs.

thomas-fossati reviewed (commented):
A really nice work. I have left a few flyby comments for your consideration.

Commit:
Update signature-envelope-cose.md

Co-authored-by: Thomas Fossati <tho.ietf@gmail.com>
Signed-off-by: David Tesar <david.tesar@microsoft.com>

Commit:
Update signature-envelope-cose.md

Co-authored-by: Thomas Fossati <tho.ietf@gmail.com>
Signed-off-by: David Tesar <david.tesar@microsoft.com>

yizha1 committed:
spec: update data type of signingTime to data/time

Signed-off-by: Yi Zha <yizha1@microsoft.com>

yizha1 committed:
Revert "spec: update data type of signingTime to data/time"

This reverts commit b5b27d463633b31848b4668c527ff5f6062eeb42.

Signed-off-by: Yi Zha <yizha1@microsoft.com>

yizha1 committed:
spec: update data type of signingTime to date/time (#199)

Signed-off-by: David Tesar <david.tesar@microsoft.com>

gokarnm commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+
+## COSE Payload
+
+The COSE envelope contains a [Notary v2 Payload](./signature-specification.md#payload).
```
Would be good to add an example of Notary V2 payload so reader does not have to jump to generic signature spec to see one.

```suggestion
The COSE envelope contains a [Notary v2 Payload](./signature-specification.md#payload).

Example of Notary v2 payload

```jsonc
{
  "targetArtifact": {
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "digest": "sha256:73c803930ea3ba1e54bc25c2bdc53edd0284c62ed651fe7b00369da519a3c333",
    "size": 16724,
    "annotations": {
        "io.wabbit-networks.buildId": "123"  // user defined metadata
    }
  }
}
```

```

gokarnm commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+  - `io.cncf.notary.authenticSigningTime`
+  - `io.cncf.notary.expiry`
+
+Example with Signing Scheme `notary.x509`
```
```suggestion
COSE examples in this document use [CBOR diagnostic notation](https://www.rfc-editor.org/rfc/rfc7049#section-6)([examples](https://datatracker.ietf.org/doc/html/rfc8152#appendix-C)).

Example with Signing Scheme `notary.x509`
```

yizha1 replied:
@gokarnm  There is already a sentence after the example, see below, so I don't think we need another note before the example.

" Note: The above examples are represented using the [extended CBOR diagnostic notation](https://datatracker.ietf.org/doc/html/rfc8152#appendix-C)."

gokarnm replied:
Yes that helps, maybe I missed it while reviewing last version. Resolving.

gokarnm commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+The COSE envelope for Notary v2 uses the following header parameters:
+
+- [Common parameters](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)
+  - Label `1`: `alg`
```
@shizhMSFT , is `alg` a required header for COSE? We included it in JWS envelope as the JWS spec requires it. If not, can it be ommited and use signature algorithm from signing key/cert as source of truth?

shizhMSFT replied:
According to the COSE spec, `alg` header MUST be present if there is no externally supplied data.

shizhMSFT replied:
In the notary scenario, there is no externally supplied data. Therefore, `alg` is a required header.

gokarnm replied:
Could you clarify that a bit more, what does externally supplied data mean in this context? Isn't signing cert's signature alg from certificate chain enough in this case?

yizha1 replied:
The externally supplied data is explained in COSE spec, see https://www.rfc-editor.org/rfc/rfc9052.html#Extern_AAD, and Notary v2 doesn't use it. The value of `alg` is explained here https://www.rfc-editor.org/rfc/rfc9052.html#name-common-cose-header-paramete, which was copied below for your convenience:
"alg: This header parameter is used to indicate the algorithm used for the security processing. This header parameter MUST be authenticated where the ability to do so exists. This support is provided by AEAD algorithms or construction (e.g., COSE_Sign and COSE_Mac0). This authentication can be done either by placing the header parameter in the protected-header-parameters bucket or as part of the externally supplied data ([Section 4.3](https://www.rfc-editor.org/rfc/rfc9052.html#Extern_AAD)). The value is taken from the "COSE Algorithms" registry (see [[COSE.Algorithms](https://www.rfc-editor.org/rfc/rfc9052.html#COSE.Algorithms)])."

gokarnm commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+- **[`crit`](https://datatracker.ietf.org/doc/html/rfc8152#section-3.1)** (*array of int/tstr*): This REQUIRED parameter (label `2`) lists the header parameters that implementations MUST understand and process. It MUST only contain parameters apart from integer labels in the range of 0 to 8. This header MUST contain `io.cncf.notary.signingScheme` which is a required critical header, and optionall ...
+- **[`content type`](https://datatracker.ietf.org/doc/html/rfc8152#section-3.1)** (*tstr*): The REQUIRED parameter content type (label `3`) is used to declare the media type of the secured content (the payload). The supported value is `application/vnd.cncf.notary.payload.v1+json`.
+- **`io.cncf.notary.signingScheme`** (*tstr*, critical): This REQUIRED header specifies the [Notary v2 Signing Scheme](./signing-scheme.md) used by the signature. Supported values are `notary.x509` and `notary.x509.signingAuthority`.
+- **`io.cncf.notary.signingTime`** (*date/time*): This header specifies the time at which the signature was generated. This is an untrusted date/time, and therefore not used in trust decisions. Its value is an Epoch-Based Date/Time defined in [RFC 8949](https://datatracker.ietf.org/doc/html/rfc8949#section-3.4.2). This claim is REQUIRED and only valid when signing scheme is `notary.x509`.
```
@shizhMSFT  Is there a reason the RFC 3339 text based date time representation supported by COSE (tag 0) is not used in favor of epoch based representation (tag 1)? More concise representation may be one reason (~ 8 vs 20 bytes), are there other stronger reasons?

shizhMSFT replied:
COSE is based on CBOR where CBOR standards for **Concise Binary** Object Representation. Concise by default is the principle.

Oppositely, JWS is based on JSON, which is text-based not binary-based. That's why RFC3339 is preferable for JWS.

gokarnm commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+Note: `<<` and `>>` are used to notate the CBOR byte string resulting from encoding the data item.
+
+- **[`x5chain`](https://datatracker.ietf.org/doc/html/draft-ietf-cose-x509-08#section-2)** (*array of bstr*): This REQUIRED parameter (label `33` by [IANA](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)) contains the ordered list of X.509 certificate or certificate chain ([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign ...
+  - **TODO** update signature specification to allow chains in protected header
```
What does the implementation use now?

shizhMSFT replied:
Currently, it is the same as JWS that chains are from unprotected header.

See https://github.com/notaryproject/notation-core-go/blob/v0.2.0-beta.1/signature/cose/envelope.go#L218

gokarnm commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+    Sig_structure = [
+        / context / 'Signature1',
+        / body_protected / << ProtectedHeaders >>,
+        / external_aad / h'',
```
What is `external_aad` , it's not defined here.

shizhMSFT replied:
The `external_aad` is the COSE external data.

yizha1 replied:
@gokarnm  Did shiwei answer your questions?

gokarnm replied:
Some content describing this field and if/how Notary v2 uses it, and any example will help.

yizha1 replied:
This field is explained here https://www.rfc-editor.org/rfc/rfc9052.html#Extern_AAD, since Notary v2 doesn't use it, so it's empty.

gokarnm commented on signature-envelope-cose.md:
````diff
@@ -0,0 +1,197 @@
+2. Construct the `Sig_structure` for `COSE_Sign1`.
+    ```yaml
+    Sig_structure = [
+        / context / 'Signature1',
````
It'd be good to add detail about `context Signature1`, that the identifying context for COSE_Sign1 is `"Signature1"`.

shizhMSFT replied:
The context `Signature1` is fixed for `COSE_Sign1` and is defined by the COSE spec. No one can change it.

gokarnm replied:
Ya, I meant just documenting what `context Signature1` is in this section.

yizha1 replied:
@gokarnm See below my proposal for the 2nd bullet, what do you think?
from 
"Construct the `Sig_structure` for `COSE_Sign1`." 
to 
"Construct the `Sig_structure` for `COSE_Sign1`. The context string `Signature1` is used to identify the context of signatures using the `COSE_Sign1` structure."

gokarnm replied:
@yizha1 how about?

"Construct the Sig_structure for `COSE_Sign1`.  The `context` field is used to identify the context of signature,   string `Signature1` is used to identity a `COSE_Sign1` structure."

gokarnm commented on signature-envelope-cose.md:
````diff
@@ -0,0 +1,197 @@
+    ]
+    ```
+3. Encode `Sig_structure` into a CBOR object as a byte string named `ToBeSigned`.
+4. Compute the signature on the `ToBeSigned` constructed in the previous step by using the signature algorithm defined in the corresponding header element: `alg`.
````
```suggestion
4. Compute the signature on the `ToBeSigned` constructed in the previous step by using the signature algorithm of the signing key, which MUST match the corresponding protected header element `alg`.
```

gokarnm commented on signature-specification.md:
```diff
@@ -165,6 +168,127 @@ These attributes are considered unsigned with respect to the signing key that ge
 - **Timestamp signature** : An OPTIONAL counter signature which provides [authentic timestamp](#signing-time)e.g. Time Stamp Authority (TSA) generated timestamp signature. Only [RFC3161](ietf-rfc3161) compliant TimeStampToken are currently supported.
 - **Signing Agent**: An OPTIONAL claim that provides the identifier of the software (e.g. Notation) that produced the signature on behalf of the user. It is an opaque string set by the software that produces the signature. It's intended primarily for diagnostic and troubleshooting purposes, this attribute is unsigned, the verifier MUST NOT validate formatting, or fail validation based on the cont ...
 
+### Supported Signature Envelopes
```
This whole section for JWS is redundant, we moved to to ./signature-envelope-jws.md , this is probably because COSE spec was branched before we split up JWS into it's own spec.

gokarnm commented on signature-specification.md:
```diff
@@ -193,6 +317,16 @@ The signing certificate's public key algorithm and size MUST be used to determin
+- The certificate MUST abide by the following key length restrictions:
+- For RSA public key, the key length MUST be 2048 bits or higher.
+  - For ECDSA public key, the key length MUST be 256 bits or higher.
+
```
Same here, this section is redundant. It's covered by and updated version of the content from line 330.

gokarnm reviewed (changes requested):
Overall looks good, except for few specific details about COSE envelope and outdated content from merge, which should be resolved.

Commit:
Merge branch 'main' into cose-envelope

priteshbandi commented on specs/signature-specification.md:
```diff
@@ -193,6 +317,16 @@ The signing certificate's public key algorithm and size MUST be used to determin
+- If the basicConstraints extension is present, the cA field MUST be set false ([RFC-5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9)).
+- The certificate MUST abide by the following key length restrictions:
+- For RSA public key, the key length MUST be 2048 bits or higher.
+  - For ECDSA public key, the key length MUST be 256 bits or higher.
```
Already covered at https://github.com/notaryproject/notaryproject/blob/main/specs/signature-specification.md#leaf-certificates.

priteshbandi commented on specs/signature-specification.md:
```diff
@@ -165,6 +168,127 @@ These attributes are considered unsigned with respect to the signing key that ge
+1. Signing certificate MUST be a valid codesigning certificate.
+1. Only JWS JSON flattened format is supported.
+   See 'Signature Envelope' section.
+
```
These were moved to [signature-envelope-jws.md](https://github.com/notaryproject/notaryproject/blob/main/specs/signature-envelope-jws.md) as part of https://github.com/notaryproject/notaryproject/commit/9cfd6a91b1dfad48886af17edf7cc8235556f550, perhaps you need to rebase changes.

priteshbandi commented on specs/signature-specification.md:
```diff
@@ -241,7 +375,9 @@ The client implementation can use the aforementioned `mediaType` to parse the si
-For [JWS JSON serialization](./signature-envelope-jwt.md) signature envelope, versioning is achieved by the `cty` field in ProtectedHeaders.
+
+- For [JWS JSON serialization](./signature-envelope-jwt.md) signature envelope, versioning is achieved by the `cty` field in ProtectedHeaders.
+- For [COSE_Sign1_Tagged](./signature-envelope-cose.md) signature envelope, versioning is achieved by the `content type` (label: `3`) field in ProtectedHeaders.  
```
```suggestion
- For [COSE_Sign1](./signature-envelope-cose.md) signature envelope, versioning is achieved by the `content type` (label: `3`) field in ProtectedHeaders.  
```

yizha1 replied:
This should not be changed. Check this https://datatracker.ietf.org/doc/html/rfc8949#section-3.4

priteshbandi commented on specs/signature-specification.md:
```diff
@@ -79,6 +80,7 @@ This specification defines the set of signed and unsigned attributes that make u
 Notary v2 supports the following envelope formats:
 
 - [JWS](./signature-envelope-jws.md)
+- [COSE Sign1](./signature-envelope-cose.md)
```
```suggestion
- [COSE_Sign1](./signature-envelope-cose.md)
```

OR13 commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+    'io.cncf.notary.signingScheme',
+    'io.cncf.notary.expiry'
+  ],
+  / cty / 3: 'application/vnd.cncf.notary.payload.v1+json',
```
this header struct looks good to me.

OR13 commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+{
+  / x5chain / 33: [
+    << DER(leafCert) >>,
+    << DER(intermediate CACert) >>,
```
I am still a bit nervous about exactly what this means, but the approach seems reasonable... I would just want to see a complete example.

OR13 commented on signature-envelope-cose.md:
```diff
@@ -0,0 +1,197 @@
+  | RSASSA-PSS with SHA-256         | -37               |
+  | RSASSA-PSS with SHA-384         | -38               |
+  | RSASSA-PSS with SHA-512         | -39               |
+  | ECDSA on secp256r1 with SHA-256 | -7                |
```
I assume lack of EdDSA is intentional here.

OR13 commented on specs/signature-specification.md:
```diff
@@ -43,6 +43,7 @@ Signature Manifest Example
 - **`blobs`** (*array of objects*): This REQUIRED property contains collection of only one [artifact descriptor][artifact-descriptor] referencing signature envelope.
   - **`mediaType`** (*string*): This REQUIRED property contains media type of signature envelope blob. Following values are supported
     - `application/jose+json`
```
This seems like maybe not correct? are you saying that compact representation is NOT allowed? thats how i read them.

OR13 commented on specs/signature-specification.md:
````diff
@@ -165,6 +168,127 @@ These attributes are considered unsigned with respect to the signing key that ge
+
+```json
+{
+   "subject": {
````
good call, since `sub` is a string... maybe you want to make it clear that this is a "private claim", that is not part of the original RFC.

gokarnm reviewed (approved):
LGTM! @priteshbandi could you review and approve as well.

yizha1 committed:
spec: update cose and signature spec according to review comments (#204)

* spec: update according to review comments
* update to RFC for the fields of the sig_structure
* update description for field context

Signed-off-by: Yi Zha <yizha1@microsoft.com>

priteshbandi commented on specs/signature-envelope-cose.md:
```diff
@@ -0,0 +1,212 @@
+Note: `<<` and `>>` are used to notate the CBOR byte string resulting from encoding the data item.
+
+- **[`x5chain`](https://datatracker.ietf.org/doc/html/draft-ietf-cose-x509-08#section-2)** (*array of bstr*): This REQUIRED parameter (label `33` by [IANA](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)) contains the ordered list of X.509 certificate or certificate chain ([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign ...
+  - **TODO** update signature specification to allow chains in protected header
```
What's the motivation behind this? can we remove it from spec and open it as issue for discussion ?

SteveLasker replied:
@letmaik, @shizhMSFT can you comment?

shizhMSFT replied:
The motivation is that the signing key of a signer may have multiple organizations associated. How to determine which organization that the signer represents at the signing time is an issue.

Here's an example, the signer has two cert chains `Company A -> Signer` and `Company B -> Signer`.
```mermaid
graph TD;

a["CA (Company A)"] --> s["Signer Cert"]
b["CA (Company B)"] --> s
```
Consider the following scenario:
1. Signer signs the artifact and associates cert chain `Company A -> Signer` to the resulted signature. In other words, the signer signs the artifact **representing the company A**.
2. Signer publishes the artifact and the signature.
3. Adversary sees the signature and replaces the cert chain in the unprotected header with a cert chain `Company B -> Signer`.
4. Now, from the view of verifiers, the signature is signed by the Signer, **representing the company B**.

This kind of attack can be mitigated by never reusing the signing key but it is better if we can sign the signing identity in the signature itself.

@yizha1 Let's remove it from spec and open it as issue for discussion.

SteveLasker replied:
@ianjmcm, can you provide some 👀 ?

priteshbandi reviewed (approved):
LGTM with one comment

yizha1 committed:
spec: move certificate chain discussion to a new issue (#209)

* move certificate chain discussion to a new issue

Signed-off-by: Yi Zha <yizha1@microsoft.com>

yizha1 committed:
Merge branch 'main' into cose-envelope

shizhMSFT reviewed (approved):
LGTM
